### PR TITLE
[BB-3175] Add asynchronous sprint automation

### PR DESCRIPTION
**Pull request by Agrendalath**
This adds Celery tasks and utils for asynchronous sprint automation.

## Prerequisites
1. Set `.env` (you can find them [here](https://vault.opencraft.com:8200/ui/vault/secrets/secret/show/core/Sprints%20Development%20Envs)). There were some recent changes.
1. In `sprints/dashboard/automation.py:notify_about_injection`, add `print(issue, message)` above `if not settings.DEBUG`.
1. In `sprints/dashboard/tasks.py:check_tickets_ready_for_sprint_task`, add `print(f"#{cell}: {emails}, {message}")` above `if not settings.DEBUG`.
1. In `sprints/dashboard/tasks.py:ping_overcommitted_users_task`, add `print(f"#{cell}: {emails}, {message}")` above `if not settings.DEBUG`.

## Testing instructions 
1. Start the backend with `docker-compose -f local.yml up --build` (this is important because we're changing dependencies here).
1. Run migrations with `docker-compose -f local.yml exec django python manage.py migrate`.
1. Create superuser with `docker-compose -f local.yml exec django python manage.py createsuperuser`
1. Open Django shell (in a separate terminal) with `docker-compose -f local.yml exec django python manage.py shell_plus` and input the following:
    ```python
    from sprints.dashboard.automation import *
    from sprints.dashboard.tasks import *
    ```
1. Run `move_out_injections_task()` and check that the outputted comments contain [BB-3630](https://tasks.opencraft.com/browse/BB-3630).
1. Go to [BB-3630](https://tasks.opencraft.com/browse/BB-3630) and add a `injection-accepted` label there.
1. Run `move_out_injections_task()` and check that the outputted comments do not contain [BB-3630](https://tasks.opencraft.com/browse/BB-3630) anymore.
1. Run `check_tickets_ready_for_sprint_task()` and check that the output makes sense. Notice that there is a special case in [BB-3668](https://tasks.opencraft.com/browse/BB-3668), as the reporter is not a member of any cell.
1. Run `ping_overcommitted_users_task()` and check that mentioned users are overcommitted (on sprints.opencraft.com).
1. Run and check the following:
    ```python
    conn = CustomJira(
        server=settings.JIRA_SERVER,
        basic_auth=(settings.JIRA_USERNAME, settings.JIRA_PASSWORD),
        options={'agile_rest_path': GreenHopperResource.AGILE_BASE_REST_PATH}
    )
    issue = conn.issue('BB-3636')
    flag_issue(conn, issue)
    # Check that the ticket is flagged now.
    unflag_issue(conn, issue)
    # Check that the ticket is unflagged now.
    ```
1. Run `schedule_sprint_tasks_task()`, then go to [Django admin](http://localhost:8000/admin/django_celery_beat/periodictask/) and check that all four tasks have been scheduled properly for the current sprint.

### CODE REVIEW

File: sprints/dashboard/automation.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    Retrieve users that have negative time left for the next sprint (i.e. are overcommitted).
+
+    :param conn: Jira connection.
+    :return: List of overcommitted users.
+    """
+    cells = get_cells(conn)
+    result = dict[str, list[User]]()
+
+    # Generate dashboards in parallel.
+    with ThreadPool(processes=settings.MULTIPROCESSING_POOL_SIZE) as pool:
```
Comment (Agrendalath) [inline sprints/dashboard/automation.py:245]: It doesn't look that elegant, but for 3 cells it optimizes this function from 50s to 20s

File: config/settings/base.py [hunk 2]
```diff
@@ -524,6 +534,50 @@
...
+# Day of the current sprint, after which no ticket should be added to the next one.
+SPRINT_ASYNC_TICKET_CREATION_CUTOFF_DAY = env.int("SPRINT_ASYNC_TICKET_CREATION_CUTOFF", 11)
+# Day of the current sprint, after which all tickets should be ready for the next sprint.
+SPRINT_ASYNC_TICKET_FINAL_CHECK_DAY = env.int("SPRINT_ASYNC_TICKET_FINAL_CHECK_DAY", 14)
+# A Jira label, which approves the ticket added to the sprint after the ticket creation cutoff day.
+SPRINT_ASYNC_INJECTION_LABEL = env.str("SPRINT_ASYNC_INJECTION_LABEL", "injection-accepted")
+# A name of the sprint, to which "injected" tickets will be moved.
+SPRINT_ASYNC_INJECTION_SPRINT = env.str("SPRINT_ASYNC_INJECTION_SPRINT", "Stretch Goals")
+# Message included as a comment to a ticket that has been moved out of the sprint as an injection.
+SPRINT_ASYNC_INJECTION_MESSAGE = "this ticket was an injection for the next sprint, so I moved it out to "
```
Comment (farhaanbukhsh) [inline config/settings/base.py:550]:
```suggestion
SPRINT_ASYNC_INJECTION_MESSAGE = "This ticket was an injection for the next sprint, so I moved it out to "
```

**Review by farhaanbukhsh — COMMENTED** [review]
I am still testing this but since I had these comments ready I will post these and if everything goes well I can just approve it. Sorry for the delay.

**Review by farhaanbukhsh — APPROVED** [review]
👍 

I have some nits in the code else this LGTM. 

- [x] I tested this: I ran the app locally and called the functions from the shell to check if they are behaving as they intend to.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~
- [x] Includes documentation
- [ ] ~~I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.~~

File: sprints/dashboard/automation.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    :param changelog: Include ticket's history.
+    """
+    sprints = get_all_sprints(conn)
+    sprints_str = ",".join((str(s.id) for s in sprints['future']))
+    return conn.search_issues(
+        jql_str=f"Sprint IN ({sprints_str})",
+        fields=list(get_issue_fields(conn, settings.JIRA_REQUIRED_FIELDS + settings.JIRA_AUTOMATION_FIELDS).values()),
+        expand="changelog" if changelog else "",  # Retrieve history of changes for each issue.
+        maxResults=0,
+    )
```
[the remark this answers is not in the document]
Comment (Agrendalath) [reply sprints/dashboard/automation.py:64]: @farhaanbukhsh, any ideas what we could add here?

File: sprints/dashboard/tasks.py [hunk 4]
```diff
@@ -292,7 +306,113 @@ def complete_sprint_task(board_id: int) -> None:
...
+@celery_app.task(ignore_result=True)
+def check_tickets_ready_for_sprint_task() -> None:
+    """Notify team members about incomplete tickets."""
+    with connect_to_jira() as conn:
+        cell_membership = get_cell_membership(conn)
+        issues = get_next_sprint_issues(conn)
+
+        for user, incomplete_issues in group_incomplete_issues(conn, issues).items():
+            # TODO: Add more flexibility to the Mattermost library, to support a nicer format here.
+            message = settings.SPRINT_ASYNC_INCOMPLETE_TICKET_MESSAGE + str(incomplete_issues)
```
[the remark this answers is not in the document]
Comment (Agrendalath) [reply sprints/dashboard/tasks.py:388]: Side note: `f-strings` are already performing the [FORMAT_VALUE](https://docs.python.org/3/library/dis.html#opcode-FORMAT_VALUE) operation, so we don't need to call `str` here (which is btw. slower than the former) :)

**Pull request merged**